### DISCUSSION TITLE
Node hosting a stream replica performs no peer discovery after reset

### DISCUSSION THREAD

**Discussion by maestroes** [rabbitmq/rabbitmq-server#7139] [Other] [answered]
### Problem description
A node hosting a stream replica fails to rejoin its cluster after reset. There is no peer discovery attempt because the node is not considered virgin in case the `stream` directory is created before `is_virgin_node()` check.

### Proposal
Probably the `stream` directory should be added to the ignored files list in `is_virgin_node()` check: https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbit/src/rabbit_mnesia.erl#L992

### Environment
RabbitMQ 3.11.3
Erlang 25.1.2

### Steps to reproduce
1. Create a three-node cluster with a consul-based peer discovery
2. Create a stream with default settings
3. Stop and reset any node
4. Start the node

**Comment by michaelklishin** [comment]
Streams are entirely orthogonal to peer discovery. Nodes perform peer discovery when they are blank, that is, their data store directory is blank.

Unless reset does not remove stream data — which I doubt — streams cannot possibly affect peer discovery.

**Reply by maestroes** [reply]
Hi Michael, thanks for the quick reaction,

AFAICS reset is performed as expected and `RABBITMQ_MNESIA_DIR` gets empty (except `cluster_nodes.config` and `nodes_running_at_shutdown` files), but the `stream` subdirectory created at the very first steps of the node boot prevents the node from detecting itself as blank a bit later and therefore prevents it from performing a peer discovery.

Let me share some excerpts from the boot log in chronological order:

```
***
The following plugins have been enabled:
  rabbitmq_peer_discovery_common
  rabbitmq_peer_discovery_consul

set 8 plugins.
Offline change; changes will take effect at broker restart.
2023-02-01 17:35:11.210372+05:00 [warning] <0.132.0> Overriding Erlang cookie using the value set in the environment
2023-02-01 17:35:11.838633+05:00 [info] <0.236.0> Stream: __test-stream_1675254648274666860 will use /var/lib/rabbitmq/mnesia/***/stream/__test-stream_1675254648274666860 for osiris log data directory
```

```
2023-02-01 17:35:16.958524+05:00 [info] <0.229.0>  Starting RabbitMQ 3.11.3 on Erlang 25.1.2 [jit]
```

```
2023-02-01 17:35:16.960148+05:00 [info] <0.229.0>  database dir   : /var/lib/rabbitmq/mnesia/***
```

```
2023-02-01 17:35:27.618272+05:00 [info] <0.229.0> Will register with peer discovery backend rabbit_peer_discovery_consul
```

And no "Node database directory at *** is empty. Assuming we need to join an existing cluster or initialise from scratch..." log message which is expected after reset.

**Reply by mkuratczyk** [reply]
Thanks for the details. You seem to understand the issue very well but at the same time it's one of those issues that personally, I don't see how that could ever happen. :) I am unable to reproduce the problem and I don't see how the `stream` folder could be created before the Mnesia tables were synced. Can you please provide full log file and the exact steps?

**Reply by maestroes** [reply]
Hi Michal,

the exact steps are:

1. Create an empty three-node cluster with a consul peer discovery mechanism (I believe that any plugin-provided, not built-in mechanism would do as well)

2. Create a queue with `Stream` type and `test` name with no additional arguments

3. Wait for cluster to replicate the stream (3 nodes are in Online and Members lists on the stream details page)

4. Stop any node via `rabbitmqctl stop_app` (2 Online, 3 Members)

5. Reset the stopped node via `rabbitmqctl reset` (still 2 Online, 3 Members)

6. Try to bring the node back (in fact to start a new clean node with the same node name)

Voila, the node doesn't try to discover its peers and creates its own new cluster.

The interesting thing is that there is no queues in the new cluster, but the `test` stream details page in old cluster shows 3 Online, 3 Members. So the stream seems to work somehow although the nodes have ended up with two independent clusters.

As you can see from the log excerpts above, the line "Stream: *** will use *** for osiris log data directory" appears at the very beginning of boot log (most likely caused by the peers continuing to try to contact the node) and the `stream` subdirectory is created at that moment, before any other subdirectories, causing further `is_virgin_node()` check to fail, AFAIU.

Could you please try to reproduce the issue using one of the plugin-based peer discovery mechanisms (consul, etcd, k8s)? If not already done of course =)

If you still need the full log please let me know.

**Reply by mkuratczyk** [reply]
Thanks. I think using a peer-discovery plugin was the missing ingredient (it changes the order of events and makes this race conditional possible). So far I haven't been able to reproduce it but I'll look into this tomorrow again. Your suggested fix is probably correct but I'd like to fully understand the problem before applying it.

**Reply by maestroes** [reply]
Hi Michal,

have you managed to reproduce the case? Please let me know if I can be of any help

And thanks for you time

**Reply by mkuratczyk** [reply]
Hi. I've tried, but was unable to. Earlier today I asked the team how they feel about just adding the folder to ignored as you suggested. Let's see if there are any objections. My main questions is whether stream replication should just resume as you described. Perhaps that's the real issue - it's a "fresh" node that is already a stream member. But streams put replica on all nodes so maybe that's the correct, if unexpected, behaviour.

**Reply by maestroes** [reply]
Ok, I see, thanks for sharing the details

**Reply by mkuratczyk** [reply]
Your suggestion has been implemented and merged. It'll be available in the next 3.10 and 3.11 patch releases.
https://github.com/rabbitmq/rabbitmq-server/pull/7222

**Reply by maestroes** [reply]
Thanks, I've just tried 3.11.9. Peer discovery works fine.

**Reply by lukebakken** [reply]
@maestroes we appreciate you following up!

**Comment by mkuratczyk** [comment, marked as answer]
Implemented in https://github.com/rabbitmq/rabbitmq-server/pull/7222
